demux/demux_textsub.c

```diff
@@ -0,0 +1,159 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <limits.h>
+#include <math.h>
+
+#include <subrandr/subrandr.h>
+
+#include "common/common.h"
+#include "demux/packet.h"
+#include "misc/bstr.h"
+#include "options/m_config.h"
+#include "options/m_option.h"
+#include "stream/stream.h"
+#include "demux.h"
+
+#define OPT_BASE_STRUCT struct demux_textsub_opts
+struct demux_textsub_opts {
+    int probesize;
+};
+
+const struct m_sub_options demux_textsub_conf = {
+    .opts = (const m_option_t[]) {
+        {"demuxer-textsub-probesize", OPT_INT(probesize), M_RANGE(32, INT_MAX)},
+        {0}
+    },
+    .size = sizeof(struct demux_textsub_opts),
+    .defaults = &(const struct demux_textsub_opts){
+        .probesize = 128,
+    },
+    .change_flags = UPDATE_DEMUXER,
+};
+
+struct format_codec_info {
+    const char *codec;
+    const char *codec_desc;
+};
+
+static const struct format_codec_info FMT_TO_CODEC[] = {
+    [SBR_SUBTITLE_FORMAT_UNKNOWN] = {NULL, NULL},
+    [SBR_SUBTITLE_FORMAT_SRV3   ] = {"textsub/srv3", "srv3"},
+    [SBR_SUBTITLE_FORMAT_WEBVTT ] = {"textsub/vtt",  "WebVTT"},
+};
+
+struct textsub_ext {
+    const char *ext;
+    struct format_codec_info codec_info;
+};
+
+static const struct textsub_ext TEXT_FORMAT_EXTS[] = {
+    {".vtt",  FMT_TO_CODEC[SBR_SUBTITLE_FORMAT_WEBVTT]},
+    {".srv3", FMT_TO_CODEC[SBR_SUBTITLE_FORMAT_SRV3]},
+    {".ytt",  FMT_TO_CODEC[SBR_SUBTITLE_FORMAT_SRV3]},
+    {NULL}
+};
+
+struct demux_textsub_priv {
+    bstr content;
+    bool exhausted;
+};
+
+static int demux_open_textsub(struct demuxer *demuxer, enum demux_check check)
+{
+    bstr filename = bstr0(demuxer->filename);
+    struct format_codec_info codec_info = {NULL, NULL};
+
+    struct demux_textsub_opts *opts = mp_get_config_group(demuxer, demuxer->global, &demux_textsub_conf);
+
+    for (const struct textsub_ext *ext = TEXT_FORMAT_EXTS; ext->ext; ++ext) {
+        if (bstr_endswith0(filename, ext->ext))
+            codec_info = ext->codec_info;
+    }
+
+    if (!codec_info.codec) {
+        int probe_size = stream_peek(demuxer->stream, opts->probesize);
+        uint8_t *probe_buffer = demuxer->stream->buffer;
+
+        sbr_subtitle_format fmt = sbr_probe_text((const char *)probe_buffer, (size_t)probe_size);
+        if(fmt < MP_ARRAY_SIZE(FMT_TO_CODEC))
+            codec_info = FMT_TO_CODEC[fmt];
+    }
+
+    if (check != DEMUX_CHECK_REQUEST && !codec_info.codec)
+        return -1;
+
+    struct demux_textsub_priv *priv = talloc_zero(demuxer, struct demux_textsub_priv);
+
+    priv->content = stream_read_complete(demuxer->stream, priv, 64 * 1024 * 1024);
+    if (priv->content.start == NULL) {
+        talloc_free(priv);
```

Comment:
I don't think you need to free here, priv is parented to demuxer, which will be destroyed after.